src/main/flight/position.c

```diff
@@ -106,126 +106,109 @@ void calculateEstimatedAltitude()
-    // tryArm() runs code that sets GPS_FIX_HOME, and if the result is no home fix, arming is blocked
-    // a check like this is essential for GPS Rescue
-    // if an altitude hold function was enabled, the same test would be good for ensuring a reasonable altitude estimate before takeoff
-
-    // If GPS Rescue is not enabled, the user can arm without a Home fix, or indeed without a 3D fix, with GPS active
-    // While disarmed, we use the current GPS value, regardless of GPS quality
-    // Also while disarmed, we monitor for a 3D fix and at least the required number of satellites
-    // If these are achieved before arming, we lock the offset at the gpsAltitude value
-    // If we don't, we wait until we get a 3D fix, and then correct the offset using the baro value
-    // This won't be very accurate, but all we need is a 3D fix.
-    // Note that without the 3D fix, GPS trust will be zero, and on getting a 3D fix, will depend on hDOP
-
-    static float gpsAltOffsetCm = 0.0f;
-    if (ARMING_FLAG(ARMED)) {
-        altitudeToFilterCm = gpsAltCm - gpsAltOffsetCm;; // use zeroed GPS altitude while armed
-        if (!altitudeOffsetSetGPS && haveBaroAlt && haveGpsAlt) {
-            // if we get a 3D fix after arming but without an altitude offset, zero GPS to the Baro reading
-            gpsAltOffsetCm = gpsAltCm - baroAltCm; // set GPS to Baro altitude once we get a GPS fix
-            altitudeOffsetSetGPS = true;
+        // *** Zero GPS Altitude on every arm or re-arm using most recent disarmed values
+        // Note that without the 3D fix, GPS trust will be zero anyway
+        static float gpsAltOffsetCm = 0.0f;
+        if (ARMING_FLAG(ARMED)) {
+            // if GPS rescue is active, can't arm unless we have fix and required sats (unless we are ignoring fix, which is crazy)
+            // but if GPS rescue is not active, or ignoring fix, we must zero with whatever GPS altitude reading we have when we arm
+            // but what if we have no 3D fix when we arm?  Then the zero is useless.
+            if (okToZeroGps) { // had a 3D fix while disarmed, ie a zero offset is available on arming
+                altitudeRelativeCm = gpsAltCm - gpsAltOffsetCm; // forced to zero while disarmed, always offset while armed
+                gpsAltitudeZeroed = true; // indicates that gps values are available and the value to filter has been zeroed
+            } else {
+                if (haveBaroAlt && haveGpsAlt) { // armed without a 3D fix, but have Baro, and later got a 3D fix
+                    // this will only be relevant for users without GPS Rescue enabled, or GPS Rescue in ignore 3D fix mode
+                    gpsAltOffsetCm = gpsAltCm - baroAltCm;
+                    altitudeRelativeCm = gpsAltCm - gpsAltOffsetCm;
+                    okToZeroGps = true;
+                }
+            }
+        } else {
```

Comment:
```suggestion
            if (!okToZeroGps && haveBaroAlt && haveGpsAlt) { // armed without a 3D fix, but have Baro, and later got a 3D fix
                    // this will only be relevant for users without GPS Rescue enabled, or GPS Rescue in ignore 3D fix mode
                    gpsAltOffsetCm = gpsAltCm - baroAltCm;
                    okToZeroGps = true;
            }
            if (okToZeroGps) { // had a 3D fix while disarmed, ie a zero offset is available on arming
                altitudeRelativeCm = gpsAltCm - gpsAltOffsetCm; // forced to zero while disarmed, always offset while armed
                gpsAltitudeZeroed = true; // indicates that gps values are available and the value to filter has been zeroed
            }
        } else {
```

Reply:
I see, that makes sense, I've implemented it.